### PR TITLE
Add type hints for "strings" folder

### PR DESCRIPTION
**Pull request by DmytroLitvinov**
### **Describe your change:**

This PR adds type hints into strings/ folder.
Related to issue #2128 

* [ ] Add an algorithm?
* [X] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [ ] All filenames are in all lowercase characters with no spaces or dashes.
* [ ] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by DmytroLitvinov** [comment]
Friendly ping @dhruvmanila :eyes:

File: strings/aho_corasick.py [hunk 1]
```diff
@@ -1,8 +1,9 @@
 from collections import deque
+from typing import Dict, List, Optional
```
Comment (dhruvmanila) [inline strings/aho_corasick.py:2]:
```suggestion
from typing import Dict, List, Union
```

File: strings/aho_corasick.py [hunk 2]
```diff
@@ -61,7 +62,7 @@ def set_fail_transitions(self):
                     + self.adlist[self.adlist[child]["fail_state"]]["output"]
                 )
 
-    def search_in(self, string):
+    def search_in(self, string: str) -> Dict:
```
Comment (dhruvmanila) [inline strings/aho_corasick.py:65]:
```suggestion
    def search_in(self, string: str) -> Dict[str, List[int]]:
```
Please be more specific

File: strings/aho_corasick.py [hunk 3]
```diff
@@ -12,13 +13,13 @@ def __init__(self, keywords):
             self.add_keyword(keyword)
         self.set_fail_transitions()
 
-    def find_next_state(self, current_state, char):
+    def find_next_state(self, current_state: int, char: str) -> Optional[List]:
```
Comment (dhruvmanila) [inline strings/aho_corasick.py:16]:
```suggestion
    def find_next_state(self, current_state: int, char: str) -> Union[int, None]:
```
Use `Union` for multiple return types and it doesn't even return a list.

File: strings/boyer_moore_search.py [hunk 4]
```diff
@@ -58,7 +59,7 @@ def mismatch_in_text(self, currentPos):
                 return currentPos + i
         return -1
 
-    def bad_character_heuristic(self):
+    def bad_character_heuristic(self) -> List:
```
Comment (dhruvmanila) [inline strings/boyer_moore_search.py:62]: Please be more specific.

File: strings/knuth_morris_pratt.py [hunk 5]
```diff
@@ -33,7 +36,7 @@ def kmp(pattern, text):
     return False
 
 
-def get_failure_array(pattern):
+def get_failure_array(pattern: str) -> List:
```
Comment (dhruvmanila) [inline strings/knuth_morris_pratt.py:39]: More specific.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

File: strings/aho_corasick.py (hunk 3 above)
Comment (DmytroLitvinov) [reply strings/aho_corasick.py:16]: My bad

**Comment by dhruvmanila** [comment]
What happened in here lol?
You will have to make an empty commit for other checks to run as right now only Travis CI ran.

**Comment by DmytroLitvinov** [comment]
@dhruvmanila, after pushing empty commit - 2 other checks were running. The result of their running is updating DIRECTORY.md.

**Comment by DmytroLitvinov** [comment]
@dhruvmanila , seems like it is okay. 

Selection_356
Selection_357

**Review by dhruvmanila — APPROVED** [review]
Alright then it's fine. Thank you for your contribution!

**Pull request merged**